eip-0006.md

```diff
@@ -0,0 +1,259 @@
+
+Clearly writing out the inputs + outputs as well as the path conditions may feel a little verbose, however this is on purpose. One only needs to focus on the required inputs + outputs when writing off-chain code for creating a transaction for a given path. On the flip side, when writing code which will live on-chain (the contracts), one only needs to pay attention to the path conditions of a given stage.
+
+Every stage, path, and the contract itself has room for a preamble/explanation of what is going on in the given section. This allows for the spec writer to add extra comments and/or make clarifications so that it is painfully clear how the multi-stage contract is supposed to work when sharing with others.
+
+Using markdown we get the benefit of being able to hyperlink the stages in the ToC, the paths for each stage, and where each path leads to. This provides us with a reasonable interface for going through informal specifications and gaining an understanding of how the different stages are inter-linked. With both an eagle's eye view using the ToC as well as with a targeted/sequential view via following path hyperlinks, we have the ability to dig into how the parts and the whole of any given informal specification work.
+
+
+### Alternative Formatting
+
+If the list format is too crowded for inputs/outputs/conditions, the following format is available as well for more complicated contracts which have a lot of registers and/or tokens.
+
+```md
+#### Inputs 
+##### Input #1
+...
+##### Input #2
+...
+
+#### Outputs
+##### Ouput #1
+...
+##### Ouput #2
+...
+
+#### Path Conditions
+##### Condition 1
+...
+##### Condition 2
+...
+
+```
+
+This allows one to use lists within each given section, thereby allowing for much more room to go into detail about specifics. This can be useful for contracts which use several registers across several boxes in their path conditions for example.
+
+With all of that said, let's now look at an example informal specification which uses the above scheme/format to define an NFT auction contract.
+
+
+
+
+# Example NFT Auction Contract
+This contract allows a user to initiate an on-chain auction for a NFT/singleton token which accepts bids in Ergs. This informal specification is based off of the discussion/original design from [this Ergo Forum thread](https://www.ergoforum.org/t/auctions-on-ergo/122).
+
+
+## ToC
+1. [Bootstrap Stage](<#bootstrap-stage>)
+2. [Auction Stage](<#auction-stage>)
+
+
+## Bootstrap Stage
+
+This stage allows a user to initiate the auction. It requires that they provide their NFT and a box with Erg which represents the starting bid. There are no hard-coded values, registers, or path conditions because this is the bootstrap stage and thus there is no contract yet.
```

Comment:
Fair point, it is a bit confusing as I was trying to keep to current Ergo terminology.

I'm personally not a big fan of "scenario" because it is a bit vague, but I agree with "script". I think we can keep the whole "multi-stage contract" concept be labeled as a "contract", while  the code locking a box as a "script".

Reply:
I think we are raising very important foundational semantic and vocabulary questions here.
We are often quite sloppy when we use the same word "contract" to name what I think are three completely different things: 
1) _spending condition_ - is basically proposition in a box (this corresponds to the checks which guard the box spending)
2) _script_ - which is a combination of many _spending conditions_ (this corresponds to the lifecycle of a box with many spending paths, for example BuyOrder, SellOrder in DEX)
3) _scenario_ - is the whole process in which many contracts may interact (this is where the whole DEX belongs)

I don't like the word "contract" because it is highly overloaded and abused in Ergo. 
In Ethereum the word "contract" have a single meaning and the meaning is always clear. 
In Ergo we have the 3 above meanings, and I would rather avoid using "contract" entirely.

I see the value of this EIP is to establish such a semantically unambiguous vocabulary so that all other contracts descriptions can become more meaningful for an external readers and newcomers. 

While we call it "informal" we still should make it unambiguous on semantic level.
That is my expectation from this EIP.

Reply:
Fair point. However we do live in the realm of blockchain, and straying away from using "contracts" because it may have been abused previously seems like a negative to the ecosystem as a whole as we're just making the barrier of entry higher for people to relate to concepts they currently understand.

Right, so with your definitions in this document currently:
- spending condition = path condition
- script = script
- scenario = contract

So a contract contains one or more scripts which contains one or more path conditions. Spending conditions can be a synonym for path conditions and used interchangeably. I think this is overall quite simple to explain to people.

I am very much against the use of "scenario" because the English connotation when explaining it to people sounds weird/very ad-hoc. People expect to write smart contacts, not scenarios. If we throw away the term "contract" and replace it with "scenario", marketing the capabilities of Ergo is going to also be extremely unintuitive and confusing.